execution-core/src/transfer/transaction.rs

```diff
@@ -0,0 +1,400 @@
+    #[must_use]
+    pub fn to_var_bytes(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+
+        let payload_bytes = self.payload.to_var_bytes();
+        bytes.extend((payload_bytes.len() as u64).to_bytes());
+        bytes.extend(payload_bytes);
+
+        bytes.extend(self.payload_hash.to_bytes());
+
+        bytes.extend((self.proof.len() as u64).to_bytes());
+        bytes.extend(&self.proof);
+
+        bytes
+    }
+
+    /// Deserialize the Transaction from a bytes buffer.
+    ///
+    /// # Errors
+    /// Errors when the bytes are not cannonical.
+    pub fn from_slice(buf: &[u8]) -> Result<Self, BytesError> {
+        let mut buf = buf;
+
+        let payload_len = usize::try_from(u64::from_reader(&mut buf)?)
+            .map_err(|_| BytesError::InvalidData)?;
+        let payload = Payload::from_slice(&buf[..payload_len])?;
+        buf = &buf[payload_len..];
+
+        let payload_hash = BlsScalar::from_reader(&mut buf)?;
+
+        let proof_len = usize::try_from(u64::from_reader(&mut buf)?)
+            .map_err(|_| BytesError::InvalidData)?;
+        let proof = buf[..proof_len].into();
+
+        Ok(Self {
+            payload,
+            payload_hash,
+            proof,
+        })
+    }
+
+    /// Return input bytes to hash the Transaction.
+    ///
+    /// Note: The result of this function is *only* meant to be used as an input
+    /// for hashing and *cannot* be used to deserialize the `Transaction` again.
+    #[must_use]
+    pub fn to_hash_input_bytes(&self) -> Vec<u8> {
+        let mut bytes = self.payload.to_hash_input_bytes();
+        bytes.extend(self.payload_hash.to_bytes());
+        bytes.extend(&self.proof);
+
+        bytes
+    }
+
+    /// Create the `Transaction`-hash.
+    #[must_use]
+    pub fn hash(&self) -> BlsScalar {
+        BlsScalar::hash_to_scalar(&self.to_hash_input_bytes())
+    }
+
+    /// Check if the `Transaction` is sensible.
+    ///
+    /// This ensures that a deposit to a contract is only possible if the
+    /// deposit value in the transaction skeleton is positive and there is an
+    /// inter-contract-call present.
+    #[must_use]
+    pub fn is_valid(&self) -> bool {
+        // make sure the payload-hash is calculated correctly
+        if self.payload().hash() != *self.payload_hash() {
+            return false;
+        }
+
+        // if the transaction is a contract deposit, the deposit value in the
+        // tx-skeleton needs to be positive and there needs to be a contract
+        // call present
+        if self.payload.deposit
+            && (self.payload.tx_skeleton.deposit == 0
+                || self.payload.contract_call.is_none())
+        {
+            return false;
+        }
+
+        // if the transaction is not a deposit, the deposit value in the
+        // transaction needs to be zero.
+        if !self.payload.deposit && self.payload.tx_skeleton.deposit > 0 {
+            return false;
+        }
+
```

Comment:
since `payload.deposit` and `payload.tx_skeleton.deposit != 0` contain exactly the same bit of information, does it make sense to have such redundancy only to check the consistency in this redundancy in this `is_valid` method. I'd consider removing the `payload.deposit` boolean.

Reply:
nitpick: I would also introduce brackets here for a less confusing operand @moCello